// zone.js/lib/wtf.js
// Detect and setup WTF.
var wtfTrace = null;
var wtfEvents = null;
var wtfEnabled = (function () {
    var wtf = global['wtf'];
    if (wtf) {
        wtfTrace = wtf['trace'];
        if (wtfTrace) {
            wtfEvents = wtfTrace['events'];
            return true;
        }
    }
    return false;
})();
function noop() {
}
exports.enabled = wtfEnabled;
exports.createScope = wtfEnabled ? function (signature, flags) {
    return wtfEvents.createScope(signature, flags);
} : function (s, f) {
    return noop;
};
exports.createEvent = wtfEnabled ? function (signature, flags) {
    return wtfEvents.createInstance(signature, flags);
} : function (s, f) {
    return noop;
};
exports.leaveScope = wtfEnabled ? function (scope, returnValue) {
    wtfTrace.leaveScope(scope, returnValue);
    return returnValue;
} : function (s, v) {
    return v;
};
exports.beginTimeRange = wtfEnabled ? function (rangeType, action) {
    return wtfTrace.beginTimeRange(rangeType, action);
} : function (t, a) {
    return null;
};
exports.endTimeRange = wtfEnabled ? function (range) {
    wtfTrace.endTimeRange(range);
} : function (r) {
};
